flent/runners.py

```diff
@@ -950,6 +956,126 @@ def find_binary(cls, host, interval, length, ip_version, local_bind=None,
+                 "FIN-WAIT-2", "TIME-WAIT", "UNCONN", "CLOSE-WAIT", "LAST-ACK",
+                 "LISTEN", "CLOSING"]
+    ss_states_re_pat = "|".join(ss_states)
+
+    def __init__(self, **kwargs):
+        super(SsRunner, self).__init__(**kwargs)
+        self.pid = str(self._parent.pid)
+
+    def filter_np_parent(self, part):
+        sub_part = []
+        sub_parts = re.split(self.ss_states_re_pat, part)
+        sub_parts = [sp for sp in sub_parts if sp.strip()]
+
+        for sp in sub_parts:
+            pid = self.pid_re.search(part)
+            if None is pid:
+                continue
+            pid_str = pid.group('pid')
+            if self.pid == pid_str:
+                sub_part.append(sp)
+
+        if 1 != len(sub_part):
+            raise ParseException()
+
+        return sub_part[0]
+
+    def parse_part(self, part):
+
+        sub_part = self.filter_np_parent(part)
+
+        timestamp = self.time_re.search(sub_part)
+        cwnd = self.cwnd_re.search(sub_part)
+        rtt = self.rtt_re.search(sub_part)
+
+        f_ports = self.ports_ipv4_re.search(sub_part)
+        if None is f_ports:
+            f_ports = self.ports_ipv6_re.search(sub_part)
+
+        if None in [timestamp, cwnd, rtt, f_ports]:
+            raise ParseException()
+
+        timestamp = float(timestamp.group('timestamp'))
+        cwnd = int(cwnd.group('cwnd'))
+        rtt, rtt_var = [float(x) for x in rtt.group('rtt').split('/')]
+        self.src_p.append(f_ports.group('src_p').split(":")[-1])
+        self.dst_p.append(f_ports.group('dst_p').split(":")[-1])
+
+        raw_values = {'t': timestamp, 'cwnd': cwnd, 'rtt': rtt, 'rtt_var': rtt_var}
+        self.raw_values.append(raw_values)
+
+        return raw_values
+
+    def parse(self, output):
+        results = {}
+        parts = output.split("\n---\n")
+        for part in parts:
+            try:
+                res_dict = self.parse_part(part)
+                t = res_dict['t']
+                for k, v in res_dict.items():
+                    if k == 't':
+                        continue
+                    if k not in results:
+                        results[k] = [[t, v]]
+                    else:
+                        results[k].append([t, v])
+
+            except ParseException:
+                continue
+
+        return results
+
+    def find_binary(self, ip_version, host, interval, length, target):
+        script = os.path.join(DATA_DIR, 'scripts', 'ss_iterate.sh')
+        if not os.path.exists(script):
+            raise RuntimeError("Cannot find ss_iterate.sh.")
+
+        bash = util.which('bash')
+        if not bash:
+            raise RuntimeError("Socket stats requires a Bash shell.")
+
+        resol_target = util.lookup_host(target, ip_version)[4][0]
+        if ip_version == 6:
+            resol_target = "[" + str(resol_target) + "]"
+
+        return "{bash} {script} -I {interval:.2f} " \
+            "-c {count:.0f} -H {host} -t {target}".format(
```

Comment:
You'll need to single-quote target, otherwise the shell is going to barf on the []